engine/resources/pippet_run.go

```diff
@@ -0,0 +1,160 @@
+// Mgmt
+// Copyright (C) 2013-2018+ James Shubin and the project contributors
+// Written by James Shubin <james@shubin.ca> and the project contributors
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
+package resources
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+	"os/exec"
+	"sync"
+)
+
+// PippetRunner is the interface to the wrapped "puppet yamlresource receive"
+// process that keeps running alongside mgmt. Individual resources are expected
+// to register themselves, and can use the Apply function from that point on.
+type PippetRunner interface {
+	Register(interface{}) error
+	Unregister(interface{}) error
+	Apply(string, string, string) PippetResult
+	Close() error
+}
+
+// PippetResult is the structured return value type for the PippetRunner's
+// Apply function.
+type PippetResult struct {
+	Error     bool
```

Comment:
What's the difference between Error and failed? And Exception?

My guess is you might prefer:

`(PippetResult, error)` (as mentioned above), with:

```type PippetResult struct{
   Changed bool
   AnyOtherNeededParam int
}```

Reply:
It's subtle, so I'm not sure we need it.

"Failed" is a flag that puppet can set during the sync operation. E.g., a *mount* resource could not mount the file system.

"Error" is a flag my own module sets to indicate that there was an issue with the input, e.g. a YAML parsing error, or an invalid resource type.

"Exception" is the message that describes what went wrong specifically in case Error==true.

A failed resource does not fill the Exception field. Any diagnostics will be written to stderr by Puppet itself, and we don't bother trying to catch those atm. The result is some red Puppet output mixed into mgmt's own output. This is not future-proof, but I would prefer not to over-engineer this detail right now.

Reply:
I think this is "fine" now because the Apply function was reworked. (?)